Mathlib/LinearAlgebra/IsTensorProduct.lean

```diff
@@ -0,0 +1,362 @@
+
+def IsTensorProduct.rightSmul.smul : SMul Cᵐᵒᵖ X where
+  smul r := rightSmul tmul r
+
+def IsTensorProduct.rightSmul.mulAction : MulAction Cᵐᵒᵖ X where
+  __ := rightSmul.smul tmul
+  one_smul := rightSmul.one_smul tmul
+  mul_smul := rightSmul.mul_smul tmul
+
+def IsTensorProduct.rightSmul.distribMulAction : DistribMulAction Cᵐᵒᵖ X where
+  __ := rightSmul.mulAction tmul
+  smul_zero _ := show rightSmul tmul _ _ = 0 by simp
+  smul_add a x y := show rightSmul tmul _ _ = rightSmul tmul _ _ + rightSmul tmul _ _ by simp
+
+def IsTensorProduct.rightModule : Module Cᵐᵒᵖ X where
+  __ := rightSmul.distribMulAction tmul
+  smul_add a x y := by simp
+  add_smul a b x := show rightSmul tmul _ _ = rightSmul tmul _ _ + rightSmul tmul _ _ by
+    refine IsTensorProduct.induction B tmul ?_ ?_ ?_ x
+    · intro m n; simp only [rightSmul_tmul, add_smul, tmul.apply_apply_add]
+    · intro x y hx hy
+      simp only [map_add, hx, hy]
+      abel
+    · intro x hx
+      simp only [map_neg, hx]
+      abel
+  zero_smul x := show rightSmul tmul _ _ = 0 by
+    refine IsTensorProduct.induction B tmul ?_ ?_ ?_ x
+    · intro m n; simp only [rightSmul_tmul, zero_smul, tmul.apply_apply_zero]
+    · intro x y hx hy
+      simp only [map_add, hx, hy]
+      abel
+    · intro x hx
+      simp only [map_neg, hx]
+      abel
+
+end right_mod
+
+end module_structure
+
+noncomputable section examples
+
+universe uR uS uM uN
+variable (R : Type uR) [CommRing R]
+variable (S : Type uS) [CommRing S]
+variable (M : Type uM) [AddCommGroup M] [Module R M]
+variable (N : Type uN) [AddCommGroup N] [Module R N] [Module S N] [SMulCommClass R S N]
+
+local instance : Module Rᵐᵒᵖ M := Module.compHom M ((RingHom.id _).fromOpposite mul_comm)
+local instance : Module Sᵐᵒᵖ N := Module.compHom N ((RingHom.id _).fromOpposite mul_comm)
+local instance : SMulCommClass R Sᵐᵒᵖ N where
+  smul_comm r s n := smul_comm r s.unop n
+
+@[simps]
+noncomputable def TensorProduct.mkAddHom : M →+ N →+ M ⊗[R] N where
+  toFun := fun m =>
+  { toFun := (m ⊗ₜ ·)
+    map_zero' := tmul_zero _ _
+    map_add' := tmul_add _ }
+  map_zero' := by aesop
+  map_add' := fun _ _ => FunLike.ext _ _ fun _ => add_tmul _ _ _
+
+noncomputable instance isTensorProduct1 : IsTensorProduct R
+  (⟨TensorProduct.mkAddHom R M N, fun r m n => smul_tmul r m n⟩ :
+    BalancedBiAddMonoidHom R M N (M ⊗[R] N)) where
+  lift P _ f := TensorProduct.liftAddHom f.toAddMonoidHom f.balanced
+  lift_comp P _ f m n := rfl
+  lift_uniq P _ f fHat H := FunLike.ext _ _ fun x => x.induction_on (by aesop)
+    (fun m n => by rw [show fHat (m ⊗ₜ n) = _ from H m n]; rfl)
+    (fun z z' h h' => by simp only [map_add, h, h'])
+
+instance : Module S (M ⊗[R] N) :=
+  have i1 := IsTensorProduct.rightModule (B := R) (C := S) (M := M) (N := N) (X := M ⊗[R] N)
+  @Module.compHom _ _ (M ⊗[R] N) _ _ (@i1 _ (isTensorProduct1 R M N)) _
+    ((RingHom.id S).toOpposite mul_comm)
+
+example (s : S) (m : M) (n : N) : s • (m ⊗ₜ[R] n) = m ⊗ₜ (s • n) := rfl
```

Comment:
when $R, S$ are both commutative rings and $M$ an $R$-module and $N$ an $(R, S)$-bimodule, then $M\otimes_R N$ is both an $R$-module and an $S$-module. this characteristic predicate can handle it